### PR TITLE
[AdaptiveTree] Rework to autodetect manifest type

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

Time ago i have started to do this rework to allow ISA autodetect the manifest type this will allow in future to full remove the current mandatory property `inputstream.adaptive.manifest_type`

The changes has been based by taking inspiration from exoplayer so detection should have a high success rate

Summary changes:
* Autodetect manifest type
* `open` method on parsers now just "open" the data for parsing, then download methods has been completely removed from adaptive tree and moved as helper method in curl utils
* As related code, this also fix problem where the representation chooser speed is updated also when we download manifest updates during playback or when HLS parser use AdaptiveTree::Download to download KID data
* As related code, has been deprecated the hack/custom `$START_NUMBER$` placeholder in manifest url, used for DASH only seem for youtube addon only, the `manifest_update_parameter` kodi property do the same, so there no reasons to kept both ways, we simplify sourcecode, addons have a very little change to do (ps. this placeholder support was not documented in wiki)

SIDENOTE:
- the `CDashTree::RefreshLiveSegments` method looks a lot changed due to indentation change, my changes are limited until to `updateTree->Open` call
- with `inputstream.adaptive.manifest_update_parameter` its possible, for now, set parameters to the url for manifest update request, this is a temporary way for a followup PR, the goal is decouple `manifest_update_parameter` with an appropriate property to set parameters for the manifest update, and a property to force enable the manifest update, i would like in future remove the "full" behaviour parameter, we should fix our broken Live dash use cases

## Motivation and context

Detect the manifest type should be automatic and not a task of addons

## How has this been tested?

tested STRM with dash, hls, and ism with utf16
tested with addon that use proxy to manipulate manifest
tested with dash on am@zon vod

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the Wiki documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: src/parser/HLSTree.h [hunk 1]
```diff
@@ -52,6 +57,22 @@ class ATTR_DLL_LOCAL CHLSTree : public AdaptiveTree
                                PLAYLIST::StreamType type) override;
 
 protected:
+  /*!
+   * \brief Download KID from media initialization section, overridable method for test project
+   */
+  virtual bool DownloadKID(std::string_view url,
```
Comment (glennguy) [inline src/parser/HLSTree.h]: I don't know if DownloadKID (Key ID) is the correct name for downloading AES keys, maybe DownloadKey?

File: src/common/AdaptiveTreeFactory.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+#include "../parser/HLSTree.h"
+#include "../parser/SmoothTree.h"
+
+using namespace PLAYLIST;
+using namespace UTILS;
+
+namespace
+{
+
+PROPERTIES::ManifestType InferManifestType(std::string_view url,
```
Comment (glennguy) [inline src/common/AdaptiveTreeFactory.cpp]: Just a comment for later - we should introduce some tests for this.

**Review by glennguy — CHANGES_REQUESTED** [review]

File: src/common/AdaptiveTreeFactory.cpp (hunk 2 above)
Comment (CastagnaIT) [reply src/common/AdaptiveTreeFactory.cpp]: i will try take a look to include it here

File: src/parser/HLSTree.h (hunk 1 above)
Comment (CastagnaIT) [reply src/parser/HLSTree.h]: done

File: src/common/AdaptiveTreeFactory.cpp (hunk 2 above)
Comment (CastagnaIT) [reply src/common/AdaptiveTreeFactory.cpp]: tests added

Comment (glennguy) [reply src/common/AdaptiveTreeFactory.cpp]: Awesome, thanks!

**Pull request merged**